File: BattleshipCS/AIPlayer.cs
namespace BattleshipCS;

public class AIPlayer : Player
{
    public const int MAX_ATTEMPTS = 100;

    private (int, int) lastHit = (-1, -1);
    private readonly List<(int, int)> potentialTargets = new();
    private readonly List<(int, int)> allPossibleMoves = new();
    private readonly List<int> shipSizes;

    public AIPlayer(string name, int boardSize) : base(name, boardSize)
    {
        // Генерируем все возможные ходы
        for (int i = 0; i < boardSize; i++)
        {
            for (int j = 0; j < boardSize; j++)
            {
                allPossibleMoves.Add((i, j));
            }
        }

        // Перемешиваем ходы
        var random = new Random();
        allPossibleMoves = allPossibleMoves.OrderBy(x => random.Next()).ToList();

        shipSizes = GameBoard.MakeShipSizes(GameBoard.DEFAULT_SHIP_CONFIG);
    }

    public override void PlaceShips()
    {
        var random = new Random();

        foreach (int size in shipSizes)
        {
            bool placed = false;
            int attempts = 0;

            while (!placed && attempts < MAX_ATTEMPTS)
            {
                int row = random.Next(MyBoard.Size);
                int col = random.Next(MyBoard.Size);
                bool horizontal = random.Next(2) == 0;

                var ship = new Ship(size, (row, col), horizontal);
                placed = MyBoard.PlaceShip(ship);
                attempts++;
            }

            if (!placed)
            {
                placed = PlaceShipAlternative(size, random);
            }
        }
    }

    private bool PlaceShipAlternative(int size, Random random)
    {
        for (int attempt = 0; attempt < MAX_ATTEMPTS; attempt++)
        {
            foreach (bool horizontal in new[] { true, false })
            {
                for (int row = 0; row < MyBoard.Size; row++)
                {
                    for (int col = 0; col < MyBoard.Size; col++)
                    {
                        var ship = new Ship(size, (row, col), horizontal);
                        if (MyBoard.PlaceShip(ship))
                        {
                            return true;
                        }
                    }
                }
            }
        }
        return false;
    }

    public override (int, int) MakeMove()
    {
        // Если есть потенциальные цели, стреляем в них
        if (potentialTargets.Count > 0)
        {
            var target = potentialTargets[^1];
            potentialTargets.RemoveAt(potentialTargets.Count - 1);
            return target;
        }

        // Иначе стреляем в случайную клетку из доступных
        if (allPossibleMoves.Count > 0)
        {
            var move = allPossibleMoves[^1];
            allPossibleMoves.RemoveAt(allPossibleMoves.Count - 1);
            return move;
        }

        // Запасной вариант
        return (0, 0);
    }

    public void UpdateAIState(Ship.ShotResult result, (int, int) coord)
    {
        if (result == Ship.ShotResult.Hit)
        {
            lastHit = coord;

            // Добавляем соседние клетки как потенциальные цели
            (int, int)[] directions = { (0, 1), (1, 0), (0, -1), (-1, 0) };
            foreach (var dir in directions)
            {
                int newRow = coord.Item1 + dir.Item1;
                int newCol = coord.Item2 + dir.Item2;

                if (newRow >= 0 && newRow < MyBoard.Size &&
                    newCol >= 0 && newCol < MyBoard.Size)
                {
                    var newTarget = (newRow, newCol);

                    // Проверяем, что этот ход еще возможен
                    if (allPossibleMoves.Contains(newTarget))
                    {
                        potentialTargets.Add(newTarget);
                        allPossibleMoves.Remove(newTarget);
                    }
                }
            }
        }
        else if (result == Ship.ShotResult.Sunk)
        {
            // Очищаем потенциальные цели при потоплении корабля
            potentialTargets.Clear();
            lastHit = (-1, -1);
        }
    }
}

File: Battleship/AIPlayer.cpp
#include "AIPlayer.hpp"
#include <random>
#include <ctime>

AIPlayer::AIPlayer(std::string name, int boardSize)
	: Player(name, boardSize)  
	, m_lastHit({ -1, -1 })
{
	// Генерируем все возможные ходы
	for (int i = 0; i < boardSize; i++)
	{
		for (int j = 0; j < boardSize; j++)
		{
			m_allPossibleMoves.push_back({ i, j });
		}
	}

	// Перемешиваем ходы
	std::random_device rd;
	std::mt19937 gen(rd());
	std::shuffle(m_allPossibleMoves.begin(), m_allPossibleMoves.end(), gen);

	shipSizes = GameBoard::MakeShipSizes(GameBoard::DEFAULT_SHIP_CONFIG);
}

// Конструктор копирования с вызовом базового класса
AIPlayer::AIPlayer(const AIPlayer& other)
	: Player(other)  // Вызов конструктора копирования базового класса
	, m_lastHit(other.m_lastHit)
	, m_potentialTargets(other.m_potentialTargets)
	, m_allPossibleMoves(other.m_allPossibleMoves)
	, shipSizes(other.shipSizes)
{
}

// Перегрузка оператора присваивания
AIPlayer& AIPlayer::operator=(const AIPlayer& other)
{
	if (this != &other) {
		Player::operator=(other);  // Вызов оператора присваивания базового класса
		m_lastHit = other.m_lastHit;
		m_potentialTargets = other.m_potentialTargets;
		m_allPossibleMoves = other.m_allPossibleMoves;
		shipSizes = other.shipSizes;
	}
	return *this;
}

// Метод для клонирования
AIPlayer* AIPlayer::Clone() const
{
	return new AIPlayer(*this);
}

void AIPlayer::PlaceShips()
{

	std::random_device rd;
	std::mt19937 gen(rd());

	for (int size : shipSizes)
	{
		bool placed = false;
		int attempts = 0;

		while (!placed && attempts < MAX_ATEMPTS) // Ограничение на попытки
		{
			int row = gen() % m_myBoard.GetSize();
			int col = gen() % m_myBoard.GetSize();
			bool horizontal = gen() % 2 == 0;

			Ship ship(size, { row, col }, horizontal);
			placed = m_myBoard.PlaceShip(ship);
			attempts++;
		}

		if (!placed)
		{
			// Если не удалось разместить, пробуем альтернативный алгоритм
			placed = PlaceShipAlternative(size, gen);
		}
	}
}

bool AIPlayer::PlaceShipAlternative(int size, std::mt19937& gen)
{
	// Альтернативный алгоритм размещения корабля
	for (int attempt = 0; attempt < MAX_ATEMPTS; attempt++)
	{
		for (bool horizontal : {true, false})
		{
			for (int row = 0; row < m_myBoard.GetSize(); row++)
			{
				for (int col = 0; col < m_myBoard.GetSize(); col++)
				{
					Ship ship(size, { row, col }, horizontal);
					if (m_myBoard.PlaceShip(ship))
					{
						return true;
					}
				}
			}
		}
	}
	return false;
}

Player::MoveType AIPlayer::MakeMove()
{
	// Если есть потенциальные цели, стреляем в них
	if (!m_potentialTargets.empty())
	{
		MoveType target = m_potentialTargets.back();
		m_potentialTargets.pop_back();
		return target;
	}

	// Иначе стреляем в случайную клетку из доступных
	if (!m_allPossibleMoves.empty())
	{
		MoveType move = m_allPossibleMoves.back();
		m_allPossibleMoves.pop_back();
		return move;
	}

	// Запасной вариант
	return { 0, 0 };
}

void AIPlayer::UpdateAIState(Ship::ShotResult result, MoveType coord)
{
	if (result == Ship::ShotResult::eHit)
	{
		m_lastHit = coord;

		// Добавляем соседние клетки как потенциальные цели
		int directions[4][2] = { {0, 1}, {1, 0}, {0, -1}, {-1, 0} };
		for (auto& dir : directions)
		{
			int newRow = coord.first + dir[0];
			int newCol = coord.second + dir[1];

			if (newRow >= 0 && newRow < m_myBoard.GetSize() &&
				newCol >= 0 && newCol < m_myBoard.GetSize())
			{
				MoveType newTarget = { newRow, newCol };

				// Проверяем, что этот ход еще возможен
				auto it = std::find(m_allPossibleMoves.begin(), m_allPossibleMoves.end(), newTarget);
				if (it != m_allPossibleMoves.end())
				{
					m_potentialTargets.push_back(newTarget);
					m_allPossibleMoves.erase(it);
				}
			}
		}
	}
	else if (result == Ship::ShotResult::eSunk)
	{
		// Очищаем потенциальные цели при потоплении корабля
		m_potentialTargets.clear();
		m_lastHit = { -1, -1 };
	}
}